coda-rag/py/Coda-RAG.ipynb

```diff
@@ -0,0 +1,539 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import asyncio\n",
+    "import os\n",
+    "import re\n",
+    "import time\n",
+    "\n",
+    "import autoevals\n",
+    "import braintrust\n",
+    "import markdownify\n",
+    "import openai\n",
+    "import requests\n",
+    "\n",
+    "\n",
+    "data = requests.get(\n",
+    "    \"https://gist.githubusercontent.com/wong-codaio/b8ea0e087f800971ca5ec9eef617273e/raw/39f8bd2ebdecee485021e20f2c1d40fd649a4c77/articles.json\"\n",
+    ").json()\n",
+    "\n",
+    "markdown_docs = [{\"id\": row[\"id\"], \"markdown\": markdownify.markdownify(row[\"body\"])} for row in data]\n",
+    "\n",
+    "i = 0\n",
+    "markdown_sections = []\n",
+    "for markdown_doc in markdown_docs:\n",
+    "    sections = re.split(r\"(.*\\n=+\\n)\", markdown_doc[\"markdown\"])\n",
+    "    current_section = \"\"\n",
+    "    for section in sections:\n",
+    "        if not section.strip():\n",
+    "            continue\n",
+    "\n",
+    "        if re.match(r\".*\\n=+\\n\", section):\n",
+    "            current_section = section\n",
+    "        else:\n",
+    "            section = current_section + section\n",
+    "            markdown_sections.append({\"doc_id\": markdown_doc[\"id\"], \"section_id\": i, \"markdown\": section.strip()})\n",
+    "            current_section = \"\"\n",
+    "            i += 1\n",
+    "\n",
+    "print(f\"Downloaded {len(markdown_sections)} Markdown sections. Here are the first 3:\")\n",
+    "markdown_sections[:3]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Use the Braintrust AI proxy to access the OpenAI API\n",
+    "\n",
+    "The [Braintrust AI proxy](https://www.braintrustdata.com/docs/guides/proxy) provides a single API to access OpenAI and Anthropic models, LLaMa 2, Mistral and others. Here we use it to access `gpt-3.5-turbo`. Because the Braintrust AI proxy automatically caches and reuses results when possible (`temperature=0` or `seed` parameter set), we can re-evaluate the following prompts many times without incurring additional API costs."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "client = braintrust.wrap_openai(openai.AsyncOpenAI(\n",
+    "    base_url=\"https://braintrustproxy.com/v1\",\n",
+    "    default_headers={\"x-bt-use-cache\": \"always\"},\n",
+    "    api_key=os.environ[\"OPENAI_API_KEY\"])\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's set up a simple wrapper function to run basic LLM calls while logging relevant metrics and metadata to a Braintrust experiment."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@braintrust.traced(name=\"OpenAI Completion\")\n",
+    "async def simple_completion(model, messages, temperature=0, seed=None, **kwargs):\n",
```

Comment:
I also think you can remove this, because the wrap function should take care of it

Reply:
and just call `client.chat.completions.create` throughout the notebook

Reply:
done. btw does the proxy auto-retry when hitting API rate limits? I'm asking because when I removed this wrapper (with its own retry logic), I didn't run into any rate limit issues even when circumventing the cache, but I may have just gotten lucky

Reply:
It does if you have multiple endpoints but not within an endpoint (IIRC), but the OpenAI driver now does.